lib/internal/Magento/Framework/Console/CommandList.php

```diff
@@ -27,9 +37,9 @@ public function __construct(array $commands = [])
     }
 
     /**
-     * {@inheritdoc}
+     * @inheritdoc
      */
-    public function getCommands()
+    public function getCommands(): array
```

Comment:
Please remove this return type as this change is Backward incompatible.

Reply:
Doesn't it make sense to slowly add in type hinting like this? PHP 5.3 is quite some time ago. And you can always add stronger typing to a class without breaking the interface. As I see it, nothing is breaking backward compatibility here?

Reply:
I'm not sure whether these arguments make sense. Yes, indeed somebody could create a DI plugin and return a string from the `getCommands()` method. But that will break the entire Magento CLI anyway! The same way counts for rewriting the `CommandList` class or `CommandListInterface` interface. So, the choice is between opting for the possibility that somebody has hacked the Magento core to an extend where the CLI does not work properly anymore, or sticking to modern-day PHP standards. I would opt for the future instead.

Reply:
> Calls getCommands() and expects a return type other than an array.
It was always returning an erray, so no code can expect it

>  Extends or implements the CommandList class or CommandListInterface without updating overridden/implemented methods.

Technically you’re right, but if we’ll treat all changes as this - we can’t add return types to anywhere --> we’ll never have strict types

Since this class was previously not marked as API, we can still change its interface.
But after merging this PR, we can’t. So I suggest adding the return type now

Reply:
Hello @ihor-sviziev & @jissereitsma,

Let me discuss this with the internal team.

Thanks

Reply:
Hello @ihor-sviziev & @jissereitsma,

We got reply from the internal team, according to them adhering to BIC checks should remain the standard unless there is a highly compelling justification to deviate from it. In this instance, I don't find the justification sufficiently convincing.